### PR TITLE
feat: database wrappers

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
The idea is to restrict all sql access to the `storage` crate. The next step would be to remove `rusqlite` as a dependency from all other crates, forcing them to use the `storage` types instead.

This is similar to what we already (partially) have, with the `XXXTable` types, except we never restricted `rusqlite` so it was often easier to just use raw sql.

The end-goal is to enable an API like:
```
impl Transaction {
  fn insert_sierra_class(&self, ...) -> anyhow::Result<()> {..}  
  fn insert_cairo_class(&self, ...) -> anyhow::Result<()> {..}
  fn insert_block(&self, ...) ...

  fn get_sierra_class(&self, class_hash) -> anyhow::Result<..> {..}
}
```

The majority of cases have been transformed to the above format. Still left todo:

- [x] some testing of the sql functions
- [x] rpc still invokes some rusqlite directly (nonce for example)
- [x] common still contains to/from sql impls, which will be removed right at the end
- [x] `storage::Transaction` still derefs to `rusqlite::Transaction`, which will be removed once all rusqlite is contained to `storage`

Something nice is also the block header type with a nice builder for easier test setup. I think we can expand similar things for state update, block body which we should then use for testing in general.

There are some notable changes which could be contenious and therefore very open to being veto'd.

## Database schism

Old database support dropped. These were really annoying to refactor as they partially contain production code so I just dumped all migrations older than 3 months. Older nodes will get greeted with an error message instead of a succesful migration.  More specifically all migrations `<= 30` are deleted and replaced with a `base_schema` migration instead. I generated the `base_schema` by applying those 30 migrations and then `.schema` it. I think this is enough? 👀 here would be good..

## Trie storage

Refactored back to being a reference counting implementation. I have ideas here for the future to make it even simpler by using just a single `trie_nodes` table instead of the several `tree_xxx` tables we have now, but that requires an expensive migration and should probably wait until we have additional logic for only persisting specific block's trie (as this would require a similar migration).

## Crate coupling

The crate coupling is a little awkward in general. A solution is to move any and all common types to `common` so that we don't have `storage -> merkle trie -> gateway` etc dependency chains.

The Trie Storage change for example inverts the current dep graph instead, and also moves `TrieNode` to common to make it easier.

## Breaking

This is a breaking version since we are dropping old databases. Ideally we should then also include other intended breaking changes like:
- default rpc to v0.3 (need to wait for starknet-js, eta ~1-2 weeks)
- cli config option renamings to categorise them better

### CODE REVIEW

**Comment by CHr15F0x** [comment]
I like the idea at least for the sake of removing the ability to use raw sql outside the storage crate. Additionally I like the idea that I would just be able to use the api to do what I want without the necessity to remember which part (table) of the api it was, which becomes to be a bit of a problem.

File: crates/storage/src/test_fixtures.rs
Comment (kkovaacs) [inline crates/storage/src/test_fixtures.rs]: Why not `tx.insert_state_diff()` here?

File: crates/storage/src/state.rs
Comment (kkovaacs) [inline crates/storage/src/state.rs:539]: Where do these tests go?

I think we have some tests for other storage functionality but didn't find any for events...

Comment (Mirko-von-Leipzig) [reply crates/storage/src/state.rs:539]: I'll be moving the tests across still; sorry really should have done that before opening the PR.

The events tests will go to `storage/src/events.rs`.

File: crates/merkle-tree/src/contract.rs
Comment (Mirko-von-Leipzig) [reply crates/merkle-tree/src/contract.rs:98]: Yeah its a result of using an internal statement.. I suppose we could use a `RefCell` or something similar to hide it?

Another reason it might be required is if one creates a storage adapter with an internal hash map cache - then it too would require `mut`.

Comment (Mirko-von-Leipzig) [reply crates/merkle-tree/src/contract.rs:98]: I'll see if its applicable for the database statement as well. I do think having the prepared statement is nice, but maybe its not worth it.

File: crates/storage/src/test_fixtures.rs
Comment (Mirko-von-Leipzig) [reply crates/storage/src/test_fixtures.rs]: I think I did this cleanup in a following commit already.

Comment (Mirko-von-Leipzig) [reply crates/storage/src/test_fixtures.rs]: I should probably make the interior methods `pub(super)` instead of `pub(crate)`..

Comment (kkovaacs) [reply crates/storage/src/test_fixtures.rs]: Yeah, totally possible that you've already done this -- the drawback of reviewing commit-by-commit. :)

And yeah, `pub(super)` is more reasonable I think.

File: crates/merkle-tree/src/contract.rs
Comment (Mirko-von-Leipzig) [reply crates/merkle-tree/src/contract.rs:99]: We've decided to keep it `&mut` as this is fine for our current codebase.

Alternatives were:
1. `RefCell<PreparedStatement>` which has `!Send` implications
2. Moving the statement into `get` which would still work somewhat okay since rusqlite statement cache includes the statement's sql string.

**Review by kkovaacs — APPROVED** [review]
LGTM!

File: crates/rpc/src/cairo/ext_py.rs
Comment (CHr15F0x) [inline crates/rpc/src/cairo/ext_py.rs:198]: I guess we could match repetition by swapping `ComputedClass::hash() -> &ClassHash` for
`ComputedClass::into_class_hash() -> ClassHash` and `ComputedClass::as_class_hash() -> &ClassHash`

File: crates/rpc/src/v02/method/get_class.rs
Comment (CHr15F0x) [inline crates/rpc/src/v02/method/get_class.rs:26]: nit: consistency with other methods
```suggestion
        other => (other.try_into().expect("Only pending cast should fail"), false),
```

File: crates/storage/src/lib.rs
Comment (CHr15F0x) [inline crates/storage/src/lib.rs:24]:
```suggestion
```

**Review by CHr15F0x — CHANGES_REQUESTED** [review]
LGTM % please have a look if `is_l1_accepted()` stuff has a bug or I just haven't had enough coffee.

Apart from that:
- only some minor nits,
- the amount of code was pretty huge and in the future you may want to split the PR into smaller ones. Especially that when going through such a big number of files I am inclined to assume some of the code has been moved around but without actually doing the diffing with the older files that have been removed on my own. I imagine you could have submitted such PRs (just a loose idea):
   1. remove old migrations
   2. add the block header type along with the builder
   3. turn some dependencies into workspace deps
   4. perform a bunch of renames
   5. add the new storage API without removing the former,
   6. use the new storage API in place of the old one, dump the old one

File: crates/merkle-tree/src/storage.rs
Comment (Mirko-von-Leipzig) [reply crates/merkle-tree/src/storage.rs:7]: The argument against it is that you aren't guaranteed to cache the statement then. As in, it will recreate the statement every `get`. And since this operation occurs many times it would be quite bad if it isn't cached.

File: crates/serde/src/lib.rs
Comment (Mirko-von-Leipzig) [reply crates/serde/src/lib.rs:30]: Because that's a lot more work :) I'd like to do that in the future, but right now I was just focussed on getting the stuff I needed for storage.

File: crates/merkle-tree/src/contract.rs
Comment (Mirko-von-Leipzig) [reply crates/merkle-tree/src/contract.rs:99]: I'm open to changing this back and just relying on the sqlite statement caching. This is easily the thing that annoys me most.

File: crates/serde/src/lib.rs
Comment (CHr15F0x) [reply crates/serde/src/lib.rs:30]: Sure

File: crates/rpc/src/lib.rs
Comment (CHr15F0x) [reply crates/rpc/src/lib.rs:283]: &mut stays, so resolving

File: crates/common/src/lib.rs
Comment (CHr15F0x) [reply crates/common/src/lib.rs:57]: It does simplify a lot of tiny code pieces here and there

File: crates/storage/src/connection/block.rs
Comment (Mirko-von-Leipzig) [reply crates/storage/src/connection/block.rs:238]: Correct. I think we need many more tests tbh. Will slowly add them as I can.

File: crates/rpc/src/cairo/ext_py.rs
Comment (Mirko-von-Leipzig) [reply crates/rpc/src/cairo/ext_py.rs:198]: In general I think we need a better, simpler abstraction for `ClassHash`, `SierraHash`, `CasmHash`, `CairoHash` (which doesn't exist atm). Or maybe more generally, it might be more correct if we do:
```rust
// Add this type.
struct CairoHash(Felt);

// Redefine this one
enum ClassHash {
    Cairo(CairoHash),
    Sierra(SierraHash),
}
```
but I think that's a problem for another day.

**Comment by Mirko-von-Leipzig** [comment]
I've pushed review changes and some fixes. I'll wait to rebase so that we don't lose track of the diffs post-review.

I'll just be adding more tests to storage stuff until we are ready to merge this sometime later.

File: crates/storage/src/connection/block.rs
Comment (CHr15F0x) [reply crates/storage/src/connection/block.rs:246]: Shouldn't this comment be removed?

Comment (Mirko-von-Leipzig) [reply crates/storage/src/connection/block.rs:246]: Thanks somehow I missed this again.

**Pull request merged**
